client/lib/json-patch-lib.ts

```diff
@@ -15,74 +15,68 @@ export interface AddOperation<T> extends BaseOperation {
   op: PatchOp.REMOVE
+  value?: undefined
 }
 
 export interface ReplaceOperation<T> extends BaseOperation {
   op: PatchOp.REPLACE
   value: T
 }
 
-export type Doc = { [key: string]: any }
-
 export type Operation = AddOperation<any> | RemoveOperation | ReplaceOperation<any>
 
-export function getValueByPointer (doc: any, pointer: string): any {
+export function getValueByPointer (doc: Record<string, any>, pointer: string): any {
   if (pointer === '/') return doc
-  const pathArr = pointer.split('/')
-  pathArr.shift()
+
   try {
-    return pathArr.reduce((acc, next) => acc[next], doc)
+    return pointer.split('/').slice(1).reduce((acc, next) => acc[next], doc)
   } catch (e) {
     return undefined
   }
 }
 
-export function applyOperation (doc: Doc, op: Operation): Operation | null {
-  let undo: Operation | null = null
-  const pathArr = op.path.split('/')
-  pathArr.shift()
-  pathArr.reduce((node, key, i) => {
-    if (!isObject) {
-      throw Error('patch cannot be applied.  Path contains non object')
-    }
-
-    if (i < pathArr.length - 1) {
-      // iterate node
-      return node[key]
-    }
-
-    // if last key
-    const curVal = node[key]
-    if (op.op === 'add' || op.op === 'replace') {
-      node[key] = op.value
-      if (curVal) {
-        undo = {
-          op: PatchOp.REPLACE,
-          path: op.path,
-          value: curVal,
-        }
-      } else {
-        undo = {
-          op: PatchOp.REMOVE,
-          path: op.path,
-        }
-      }
-    } else {
-      delete node[key]
-      if (curVal) {
-        undo = {
-          op: PatchOp.ADD,
-          path: op.path,
-          value: curVal,
-        }
-      }
-    }
-  }, doc)
-
-  return undo
+export function applyOperation (doc: DBCache<Record<string, any>>, { path, op, value }: Operation): Operation | null {
+  const current = getValueByPointer(doc.data, path)
+  const remove = { op: PatchOp.REMOVE, path} as const
+  const add = { op: PatchOp.ADD, path, value: current} as const
+  const replace = { op: PatchOp.REPLACE, path, value: current } as const
+
+  doc.data = recursiveApply(doc.data, path.split('/').slice(1), value)
+
+  switch (op) {
+    case PatchOp.REMOVE:
+      return current === undefined
+        ? null
+        : add
+    case PatchOp.REPLACE:
+    case PatchOp.ADD:
+      return current === undefined
+        ? remove
+        : replace
+  }
+}
+
+function recursiveApply<T extends Record<string, T>> (data: T, path: readonly string[], value?: any): T {
+  if (!path.length) return value
+
+  if (!isObject(data)) {
+    throw Error('Patch cannot be applied. Path contains non object')
+  }
+
+  const updated = recursiveApply(data[path[0]], path.slice(1), value)
+  const result = {
+    ...data,
+    [path[0]]: updated,
+  }
+
+  if (updated === undefined) {
+    delete result[path[0]]
+  }
+
+  return result
 }
 
-function isObject (val: any): val is Doc {
+function isObject (val: any): val is Record<string, any> {
```

Comment:
Instead of using `any`, use `unknown` so that potential type errors won't be hidden.

Once an `any` comes into play, type checking essentially gets turned off.